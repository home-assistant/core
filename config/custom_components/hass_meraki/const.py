"""Constants for the Meraki integration."""

DOMAIN = "hass_meraki"
