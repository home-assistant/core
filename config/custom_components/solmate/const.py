"""Constants for the solmate integration."""

DOMAIN = "solmate"
