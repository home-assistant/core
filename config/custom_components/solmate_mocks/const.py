"""Constants for the solmate_mocks integration."""

DOMAIN = "solmate_mocks"
