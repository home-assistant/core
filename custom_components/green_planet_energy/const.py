"""Constants for the Green Planet Energy integration."""

DOMAIN = "green_planet_energy"

# Config flow keys
CONF_API_URL = "api_url"

# Default values
DEFAULT_API_URL = "https://mein.green-planet-energy.de/p2"
DEFAULT_SCAN_INTERVAL = 60  # minutes

# API constants
API_METHOD = "getVerbrauchspreisUndWindsignal"
AGGREGATS_ZEITRAUM = ""
AGGREGATS_TYP = ""
SOURCE = "Portal"

# Sensor keys for each hour (9-18)
SENSOR_HOURS = [9, 10, 11, 12, 13, 14, 15, 16, 17, 18]
