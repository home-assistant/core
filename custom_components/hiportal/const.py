"""Constants for hiportal."""

DOMAIN = "hiportal"

CONF_USERNAME = "username"
CONF_PASSWORD = "password"
