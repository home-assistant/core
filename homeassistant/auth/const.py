"""Constants for the auth module."""

from datetime import timedelta

ACCESS_TOKEN_EXPIRATION = timedelta(minutes=30)
MFA_SESSION_EXPIRATION = timedelta(minutes=5)
REFRESH_TOKEN_EXPIRATION = timedelta(days=90).total_seconds()

GROUP_ID_ADMIN = "system-admin"
GROUP_ID_USER = "system-users"
GROUP_ID_READ_ONLY = "system-read-only"
