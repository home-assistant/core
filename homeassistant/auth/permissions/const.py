"""Permission constants."""

CAT_ENTITIES = "entities"
CAT_CONFIG_ENTRIES = "config_entries"
SUBCAT_ALL = "all"

POLICY_READ = "read"
POLICY_CONTROL = "control"
POLICY_EDIT = "edit"
