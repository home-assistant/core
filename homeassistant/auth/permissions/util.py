"""Helpers to deal with permissions."""

from __future__ import annotations

from collections.abc import Callable
from functools import wraps
from typing import cast

from .const import SUBCAT_ALL
from .models import PermissionLookup
from .types import CategoryType, SubCategoryDict, ValueType

type LookupFunc = Callable[[PermissionLookup, SubCategoryDict, str], ValueType | None]
type SubCatLookupType = dict[str, LookupFunc]


def lookup_all(
    perm_lookup: PermissionLookup, lookup_dict: SubCategoryDict, object_id: str
) -> ValueType:
    """Look up permission for all."""
    # In case of ALL category, lookup_dict IS the schema.
    return cast(ValueType, lookup_dict)


def compile_policy(
    policy: CategoryType, subcategories: SubCatLookupType, perm_lookup: PermissionLookup
) -> Callable[[str, str], bool]:
    """Compile policy into a function that tests policy.

    Subcategories are mapping key -> lookup function, ordered by highest
    priority first.
    """
    # None, False, empty dict
    if not policy:

        def apply_policy_deny_all(entity_id: str, key: str) -> bool:
            """Decline all."""
            return False

        return apply_policy_deny_all

    if policy is True:

        def apply_policy_allow_all(entity_id: str, key: str) -> bool:
            """Approve all."""
            return True

        return apply_policy_allow_all

    assert isinstance(policy, dict)

    funcs: list[Callable[[str, str], bool | None]] = []

    for key, lookup_func in subcategories.items():
        lookup_value = policy.get(key)

        # If any lookup value is `True`, it will always be positive
        if isinstance(lookup_value, bool):
            return lambda object_id, key: True

        if lookup_value is not None:
            funcs.append(_gen_dict_test_func(perm_lookup, lookup_func, lookup_value))

    if len(funcs) == 1:
        func = funcs[0]

        @wraps(func)
        def apply_policy_func(object_id: str, key: str) -> bool:
            """Apply a single policy function."""
            return func(object_id, key) is True

        return apply_policy_func

    def apply_policy_funcs(object_id: str, key: str) -> bool:
        """Apply several policy functions."""
        for func in funcs:
            if (result := func(object_id, key)) is not None:
                return result
        return False

    return apply_policy_funcs


def _gen_dict_test_func(
    perm_lookup: PermissionLookup, lookup_func: LookupFunc, lookup_dict: SubCategoryDict
) -> Callable[[str, str], bool | None]:
    """Generate a lookup function."""

    def test_value(object_id: str, key: str) -> bool | None:
        """Test if permission is allowed based on the keys."""
        schema: ValueType = lookup_func(perm_lookup, lookup_dict, object_id)

        if schema is None or isinstance(schema, bool):
            return schema

        assert isinstance(schema, dict)

        return schema.get(key)

    return test_value


def test_all(policy: CategoryType, key: str) -> bool:
    """Test if a policy has an ALL access for a specific key."""
    if not isinstance(policy, dict):
        return bool(policy)

    all_policy = policy.get(SUBCAT_ALL)

    if not isinstance(all_policy, dict):
        return bool(all_policy)

    return all_policy.get(key, False)  # type: ignore[no-any-return]
