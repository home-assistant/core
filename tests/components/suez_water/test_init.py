"""Test Suez_water integration initialization."""

from datetime import datetime, timedelta
from unittest.mock import AsyncMock

from freezegun.api import FrozenDateTimeFactory
import pytest
from syrupy import SnapshotAssertion

from homeassistant.components.recorder.statistics import statistics_during_period
from homeassistant.components.suez_water.const import (
    CONF_COUNTER_ID,
    DATA_REFRESH_INTERVAL,
    DOMAIN,
)
from homeassistant.components.suez_water.coordinator import (
    PySuezError,
    TelemetryMeasure,
)
from homeassistant.config_entries import ConfigEntryState
from homeassistant.const import CONF_USERNAME
from homeassistant.core import HomeAssistant
import homeassistant.util.dt as dt_util

from . import setup_integration
from .conftest import MOCK_DATA

from tests.common import MockConfigEntry, async_fire_time_changed
from tests.components.recorder.common import async_wait_recording_done


async def test_initialization_setup_api_error(
    hass: HomeAssistant,
    suez_client: AsyncMock,
    mock_config_entry: MockConfigEntry,
) -> None:
    """Test that suez_water needs to retry loading if api failed to connect."""

    suez_client.check_credentials.side_effect = PySuezError("Test failure")
    await setup_integration(hass, mock_config_entry)

    assert mock_config_entry.state is ConfigEntryState.SETUP_RETRY


async def test_init_auth_failed(
    hass: HomeAssistant,
    suez_client: AsyncMock,
    mock_config_entry: MockConfigEntry,
) -> None:
    """Test that suez_water reflect authentication failure."""
    suez_client.check_credentials.return_value = False
    await setup_integration(hass, mock_config_entry)

    assert mock_config_entry.state is ConfigEntryState.SETUP_ERROR


async def test_init_refresh_failed(
    hass: HomeAssistant,
    suez_client: AsyncMock,
    mock_config_entry: MockConfigEntry,
) -> None:
    """Test that suez_water reflect authentication failure."""
    suez_client.fetch_aggregated_data.side_effect = PySuezError("Update failed")
    await setup_integration(hass, mock_config_entry)

    assert mock_config_entry.state is ConfigEntryState.SETUP_RETRY


async def test_init_statistics_failed(
    hass: HomeAssistant,
    suez_client: AsyncMock,
    mock_config_entry: MockConfigEntry,
) -> None:
    """Test that suez_water reflect authentication failure."""
    suez_client.fetch_all_daily_data.side_effect = PySuezError("Update failed")
    await setup_integration(hass, mock_config_entry)

    assert mock_config_entry.state is ConfigEntryState.SETUP_RETRY


@pytest.mark.usefixtures("recorder_mock")
async def test_statistics_no_price(
    hass: HomeAssistant,
    suez_client: AsyncMock,
    mock_config_entry: MockConfigEntry,
) -> None:
    """Test that suez_water statistics does not register when no price."""
    # New data retrieved but no price
    suez_client.get_price.side_effect = PySuezError("will fail")
    suez_client.fetch_all_daily_data.return_value = [
        TelemetryMeasure(
            (datetime.now().date()).strftime("%Y-%m-%d %H:%M:%S"), 0.5, 0.5
        )
    ]

    await setup_integration(hass, mock_config_entry)

    assert mock_config_entry.state is ConfigEntryState.LOADED

    statistic_id = (
        f"{DOMAIN}:{mock_config_entry.data[CONF_COUNTER_ID]}_water_cost_statistics"
    )
    stats = await hass.async_add_executor_job(
        statistics_during_period,
        hass,
        datetime.now() - timedelta(days=1),
        None,
        [statistic_id],
        "hour",
        None,
        {"start", "state", "mean", "min", "max", "last_reset", "sum"},
    )

    assert stats.get(statistic_id) is None


@pytest.mark.usefixtures("recorder_mock")
@pytest.mark.parametrize(
    "statistic",
    [
        "water_cost_statistics",
        "water_consumption_statistics",
    ],
)
async def test_statistics(
    hass: HomeAssistant,
    suez_client: AsyncMock,
    mock_config_entry: MockConfigEntry,
    freezer: FrozenDateTimeFactory,
    snapshot: SnapshotAssertion,
    statistic: str,
) -> None:
    """Test that suez_water statistics are working."""
    nb_samples = 3

    start = datetime.fromisoformat("2024-12-04T02:00:00.0")
    freezer.move_to(start)

    origin = dt_util.start_of_local_day(start.date()) - timedelta(days=nb_samples)
    result = [
        TelemetryMeasure(
            date=((origin + timedelta(days=d)).date()).strftime("%Y-%m-%d %H:%M:%S"),
            volume=0.5,
            index=0.5 * (d + 1),
        )
        for d in range(nb_samples)
    ]
    suez_client.fetch_all_daily_data.return_value = result

    await setup_integration(hass, mock_config_entry)

    assert mock_config_entry.state is ConfigEntryState.LOADED

    # Init data retrieved
    await _test_for_data(
        hass,
        suez_client,
        snapshot,
        statistic,
        origin,
        mock_config_entry.data[CONF_COUNTER_ID],
        1,
    )

    # No new data retrieved
    suez_client.fetch_all_daily_data.return_value = []
    freezer.tick(DATA_REFRESH_INTERVAL)
    async_fire_time_changed(hass)

    await _test_for_data(
        hass,
        suez_client,
        snapshot,
        statistic,
        origin,
        mock_config_entry.data[CONF_COUNTER_ID],
        2,
    )
    # Old data retrieved
    suez_client.fetch_all_daily_data.return_value = [
        TelemetryMeasure(
            date=(origin.date() - timedelta(days=1)).strftime("%Y-%m-%d %H:%M:%S"),
            volume=0.5,
            index=0.5 * (121 + 1),
        )
    ]
    freezer.tick(DATA_REFRESH_INTERVAL)
    async_fire_time_changed(hass)

    await _test_for_data(
        hass,
        suez_client,
        snapshot,
        statistic,
        origin,
        mock_config_entry.data[CONF_COUNTER_ID],
        3,
    )

    # New daily data retrieved
    suez_client.fetch_all_daily_data.return_value = [
        TelemetryMeasure(
            date=(datetime.now().date()).strftime("%Y-%m-%d %H:%M:%S"),
            volume=0.5,
            index=0.5 * (121 + 1),
        )
    ]
    freezer.tick(DATA_REFRESH_INTERVAL)
    async_fire_time_changed(hass)

    await _test_for_data(
        hass,
        suez_client,
        snapshot,
        statistic,
        origin,
        mock_config_entry.data[CONF_COUNTER_ID],
        4,
    )


async def _test_for_data(
    hass: HomeAssistant,
    suez_client: AsyncMock,
    snapshot: SnapshotAssertion,
    statistic: str,
    origin: datetime,
    counter_id: str,
    nb_calls: int,
) -> None:
    await hass.async_block_till_done(True)
    await async_wait_recording_done(hass)

    assert suez_client.fetch_all_daily_data.call_count == nb_calls
    statistic_id = f"{DOMAIN}:{counter_id}_{statistic}"
    stats = await hass.async_add_executor_job(
        statistics_during_period,
        hass,
        origin - timedelta(days=1),
        None,
        [statistic_id],
        "hour",
        None,
        {"start", "state", "mean", "min", "max", "last_reset", "sum"},
    )
    assert stats == snapshot(name=f"test_statistics_call{nb_calls}")


async def test_migration_version_rollback(
    hass: HomeAssistant,
    suez_client: AsyncMock,
) -> None:
    """Test that downgrading from a future version is not possible."""
    future_entry = MockConfigEntry(
        unique_id=MOCK_DATA[CONF_COUNTER_ID],
        domain=DOMAIN,
        title="Suez mock device",
        data=MOCK_DATA,
        version=3,
    )
    await setup_integration(hass, future_entry)
    assert future_entry.state is ConfigEntryState.MIGRATION_ERROR


async def test_no_migration_current_version(
    hass: HomeAssistant,
    suez_client: AsyncMock,
) -> None:
    """Test that a current version does not migrate."""
    current_entry = MockConfigEntry(
        unique_id=MOCK_DATA[CONF_COUNTER_ID],
        domain=DOMAIN,
        title="Suez mock device",
        data=MOCK_DATA,
        version=2,
    )
    await setup_integration(hass, current_entry)
    assert current_entry.state is ConfigEntryState.LOADED
    assert current_entry.unique_id == MOCK_DATA[CONF_COUNTER_ID]


async def test_migration_version_1_to_2(
    hass: HomeAssistant,
    suez_client: AsyncMock,
) -> None:
    """Test that a migration from 1 to 2 changes the unique_id."""
    past_entry = MockConfigEntry(
        unique_id=MOCK_DATA[CONF_USERNAME],
        domain=DOMAIN,
        title=MOCK_DATA[CONF_USERNAME],
        data=MOCK_DATA,
        version=1,
    )

    await setup_integration(hass, past_entry)
    assert past_entry.state is ConfigEntryState.LOADED
    assert past_entry.unique_id == MOCK_DATA[CONF_COUNTER_ID]
    assert past_entry.title == MOCK_DATA[CONF_USERNAME]
    assert past_entry.version == 2
