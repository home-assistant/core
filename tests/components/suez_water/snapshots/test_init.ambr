# serializer version: 1
# name: test_statistics[water_consumption_statistics][test_statistics_call1]
  defaultdict({
    'suez_water:123456_water_consumption_statistics': list([
      dict({
        'end': 1733043600.0,
        'last_reset': None,
        'start': 1733040000.0,
        'state': 500.0,
        'sum': 500.0,
      }),
      dict({
        'end': 1733130000.0,
        'last_reset': None,
        'start': 1733126400.0,
        'state': 500.0,
        'sum': 1000.0,
      }),
      dict({
        'end': 1733216400.0,
        'last_reset': None,
        'start': 1733212800.0,
        'state': 500.0,
        'sum': 1500.0,
      }),
    ]),
  })
# ---
# name: test_statistics[water_consumption_statistics][test_statistics_call2]
  defaultdict({
    'suez_water:123456_water_consumption_statistics': list([
      dict({
        'end': 1733043600.0,
        'last_reset': None,
        'start': 1733040000.0,
        'state': 500.0,
        'sum': 500.0,
      }),
      dict({
        'end': 1733130000.0,
        'last_reset': None,
        'start': 1733126400.0,
        'state': 500.0,
        'sum': 1000.0,
      }),
      dict({
        'end': 1733216400.0,
        'last_reset': None,
        'start': 1733212800.0,
        'state': 500.0,
        'sum': 1500.0,
      }),
    ]),
  })
# ---
# name: test_statistics[water_consumption_statistics][test_statistics_call3]
  defaultdict({
    'suez_water:123456_water_consumption_statistics': list([
      dict({
        'end': 1733043600.0,
        'last_reset': None,
        'start': 1733040000.0,
        'state': 500.0,
        'sum': 500.0,
      }),
      dict({
        'end': 1733130000.0,
        'last_reset': None,
        'start': 1733126400.0,
        'state': 500.0,
        'sum': 1000.0,
      }),
      dict({
        'end': 1733216400.0,
        'last_reset': None,
        'start': 1733212800.0,
        'state': 500.0,
        'sum': 1500.0,
      }),
    ]),
  })
# ---
# name: test_statistics[water_consumption_statistics][test_statistics_call4]
  defaultdict({
    'suez_water:123456_water_consumption_statistics': list([
      dict({
        'end': 1733043600.0,
        'last_reset': None,
        'start': 1733040000.0,
        'state': 500.0,
        'sum': 500.0,
      }),
      dict({
        'end': 1733130000.0,
        'last_reset': None,
        'start': 1733126400.0,
        'state': 500.0,
        'sum': 1000.0,
      }),
      dict({
        'end': 1733216400.0,
        'last_reset': None,
        'start': 1733212800.0,
        'state': 500.0,
        'sum': 1500.0,
      }),
      dict({
        'end': 1733389200.0,
        'last_reset': None,
        'start': 1733385600.0,
        'state': 500.0,
        'sum': 2000.0,
      }),
    ]),
  })
# ---
# name: test_statistics[water_cost_statistics][test_statistics_call1]
  defaultdict({
    'suez_water:123456_water_cost_statistics': list([
      dict({
        'end': 1733043600.0,
        'last_reset': None,
        'start': 1733040000.0,
        'state': 2.37,
        'sum': 2.37,
      }),
      dict({
        'end': 1733130000.0,
        'last_reset': None,
        'start': 1733126400.0,
        'state': 2.37,
        'sum': 4.74,
      }),
      dict({
        'end': 1733216400.0,
        'last_reset': None,
        'start': 1733212800.0,
        'state': 2.37,
        'sum': 7.11,
      }),
    ]),
  })
# ---
# name: test_statistics[water_cost_statistics][test_statistics_call2]
  defaultdict({
    'suez_water:123456_water_cost_statistics': list([
      dict({
        'end': 1733043600.0,
        'last_reset': None,
        'start': 1733040000.0,
        'state': 2.37,
        'sum': 2.37,
      }),
      dict({
        'end': 1733130000.0,
        'last_reset': None,
        'start': 1733126400.0,
        'state': 2.37,
        'sum': 4.74,
      }),
      dict({
        'end': 1733216400.0,
        'last_reset': None,
        'start': 1733212800.0,
        'state': 2.37,
        'sum': 7.11,
      }),
    ]),
  })
# ---
# name: test_statistics[water_cost_statistics][test_statistics_call3]
  defaultdict({
    'suez_water:123456_water_cost_statistics': list([
      dict({
        'end': 1733043600.0,
        'last_reset': None,
        'start': 1733040000.0,
        'state': 2.37,
        'sum': 2.37,
      }),
      dict({
        'end': 1733130000.0,
        'last_reset': None,
        'start': 1733126400.0,
        'state': 2.37,
        'sum': 4.74,
      }),
      dict({
        'end': 1733216400.0,
        'last_reset': None,
        'start': 1733212800.0,
        'state': 2.37,
        'sum': 7.11,
      }),
    ]),
  })
# ---
# name: test_statistics[water_cost_statistics][test_statistics_call4]
  defaultdict({
    'suez_water:123456_water_cost_statistics': list([
      dict({
        'end': 1733043600.0,
        'last_reset': None,
        'start': 1733040000.0,
        'state': 2.37,
        'sum': 2.37,
      }),
      dict({
        'end': 1733130000.0,
        'last_reset': None,
        'start': 1733126400.0,
        'state': 2.37,
        'sum': 4.74,
      }),
      dict({
        'end': 1733216400.0,
        'last_reset': None,
        'start': 1733212800.0,
        'state': 2.37,
        'sum': 7.11,
      }),
      dict({
        'end': 1733389200.0,
        'last_reset': None,
        'start': 1733385600.0,
        'state': 2.37,
        'sum': 9.48,
      }),
    ]),
  })
# ---
